### ISSUE TITLE
ultralytics.utils.metrics.mask_iou has an integer overflow if you are working with uint8 masks.

### ISSUE THREAD

**Issue by djwessel** [ultralytics/ultralytics#15782]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

As a reference, here is the code for `mask_iou`

```python
def mask_iou(mask1, mask2, eps=1e-7):
    """
    Calculate masks IoU.

    Args:
        mask1 (torch.Tensor): A tensor of shape (N, n) where N is the number of ground truth objects and n is the
                        product of image width and height.
        mask2 (torch.Tensor): A tensor of shape (M, n) where M is the number of predicted objects and n is the
                        product of image width and height.
        eps (float, optional): A small value to avoid division by zero. Defaults to 1e-7.

    Returns:
        (torch.Tensor): A tensor of shape (N, M) representing masks IoU.
    """
    intersection = torch.matmul(mask1, mask2.T).clamp_(0)
    union = (mask1.sum(1)[:, None] + mask2.sum(1)[None]) - intersection  # (area1 + area2) - intersection
    return intersection / (union + eps)
```

If mask1 and mask2 have dtype of uint8 (which is rather standard for masks IMO), the output of the `torch.matmul` line is also uint8, meaning if your intersection of the masks is > 255, you end up with an integer overflow.

### Environment

`ultralytics==8.2.79`

### Minimal Reproducible Example

```python
from ultralytics.utils.metrics import mask_iou
iou = mask_iou(torch.ones(1, 10000, dtype=torch.uint8), torch.ones(1, 10000, dtype=torch.uint8))
print(iou)
```

You would expect the IoU to be 1, but instead you get 0.0008. This is because the intersect is 16 due to the integer overflow.

### Additional

To fix this, cast mask1 and mask2 to a larger data type like `torch.uint32`

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by Alarmod** [comment]
https://pytorch.org/docs/stable/tensors.html#id12

Unsigned types asides from uint8 are currently planned to only have limited support in eager mode (they primarily exist to assist usage with torch.compile); if you need eager support and the extra range is not needed, we recommend using their signed variants instead. See https://github.com/pytorch/pytorch/issues/58734 for more details.

torch.int32 :)

**Comment by glenn-jocher** [comment]
@Alarmod thank you for pointing that out. We'll update the `mask_iou` function to cast `mask1` and `mask2` to `torch.int32` to prevent integer overflow. Feel free to submit a PR with this change.
